### PR TITLE
Fix Contact Rollups connection failure after issuing multi-statement command

### PR DESCRIPTION
**Pull request by sureshc**
This change might open up the possibility of consolidating all writes to the replica server to a Temporary table.

### CODE REVIEW

**Review by wjordan — APPROVED** [review]
looks like it could be just what we need, nice work finding this! 🎉

File: lib/cdo/contact_rollups.rb [hunk 1]
```diff
@@ -808,11 +808,10 @@ def self.update_data_from_forms(form_kind)
...
 
-      # Discard the connection and get a new one after each batch. Currently, if you use multiple statements in a batch, the first call will succeed; the next call on
-      # the same connection AFTER you do a multi-statement call will fail with a MySQL error that commands are out of order. The innertubes suggest that this may be a problem in mysql2 gem.
-      # Fortunately, it is fairly cheap to get a new connection for each batch (time to get new connection is negligible compared to batch time).
-      conn.disconnect
-      conn = mysql_multi_connection
+      # https://www.rubydoc.info/github/brianmario/mysql2/Mysql2%2FClient:abandon_results!
+      # When using MULTI_STATEMENTS support, calling this will throw away any unprocessed results as fast as it can in
+      # order to put the connection back into a state where queries can be issued again.
+      conn.abandon_results!
```
Comment (sureshc) [inline lib/cdo/contact_rollups.rb]: Reading more closely, `abandon_results!` is a method on the MySQL `Client` Class.  In this part of ContactRollups, `conn` is an instance of `Sequel::Mysql2::Database`.   We may need to find a way to invoke this directly on the MySQL Client that Sequel is using.

Comment (wjordan) [reply lib/cdo/contact_rollups.rb]: Looks like the syntax would be `conn.pool.hold(&:abandon_results!)` to invoke this command directly on the MySQL client that Sequel is using.

**Review by hacodeorg — APPROVED** [review]
Thanks for finding the solution!

**Comment by sureshc** [comment]
Contact Rollups failed with this error when I tried testing out this change in production:
```
Updating ages taught for age 20                                                                                                                                                                    
   Completed in 16.663503927 seconds                                                                                                                                                               
Updating ages taught for age 21                                                                                                                                                                    
   Completed in 37.296895298 seconds                                                                                                                                                               
Updating district information from users                                                                                                                                                           
   Completed in 186.665379895 seconds                                                                                                                                                              
Updating district information from dashboard.pd_enrollments                                                                                                                                        
   Completed in 83.420391572 seconds                                                                                                                                                               
Updating school information from dashboard.pd_enrollments                                                                                                                                          
   Completed in 0.127840211 seconds                                                                                                                                                                
Updating courses_facilitated                                                                                                                                                                       
   Completed in 11.461469856 seconds                                                                                                                                                               
   Completed in 70.26900918 seconds                                                                                                                                                                
   Completed in 15.648449473 seconds                                                                                                                                                               
Appending 'hoc_organizer_years' field with '2013' from forms of kind 'CSEdWeekEvent2013'                                                                                                           
   Completed in 1.101725845 seconds                                                                                                                                                                
Appending 'hoc_organizer_years' field with '2014' from forms of kind 'HocSignup2014'                                                                                                               
   Completed in 2.326550498 seconds                                                                                                                                                                
Appending 'hoc_organizer_years' field with '2015' from forms of kind 'HocSignup2015'                                                                                                               
   Completed in 6.796637323 seconds                                                                                                                                                                
Appending 'hoc_organizer_years' field with '2016' from forms of kind 'HocSignup2016'                                                                                                               
   Completed in 6.287830217 seconds                                                                                                                                                                
Appending 'hoc_organizer_years' field with '2017' from forms of kind 'HocSignup2017'                                                                                                               
   Completed in 5.380353159 seconds                                                                                                                                                                
Appending 'hoc_organizer_years' field with '2018' from forms of kind 'HocSignup2018'                                                                                                               
   Completed in 9.64105108 seconds                                                                                                                                                                 
Appending 'roles' field with 'Petition Signer' from forms of kind 'Petition'                                                                                                                       
   Completed in 123.619605551 seconds                                                                                                                                                              
Updating data for form kind BringToSchool2013                                                                                                                                                      
Total records processed: 4600                                                                                                                                                                      
Traceback (most recent call last):                                                                                                                                                                 
       16: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/adapters/mysql2.rb:237:in `fetch_rows'                                                                                            
       15: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/adapters/mysql2.rb:274:in `execute'                                                                                               
       14: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/dataset/actions.rb:1082:in `execute'                                                                                              
       13: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/adapters/utils/mysql_mysql2.rb:38:in `execute'                                                                                    
       12: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/database/connecting.rb:270:in `synchronize'                                                                                       
       11: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/connection_pool/threaded.rb:91:in `hold'                                                                                          
       10: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/adapters/utils/mysql_mysql2.rb:38:in `block in execute'                                                                           
        9: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/adapters/mysql2.rb:152:in `_execute'                                                                                              
        8: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/adapters/mysql2.rb:239:in `block in fetch_rows'                                                                                   
        7: from /var/lib/gems/2.5.0/gems/rack-mini-profiler-1.0.0/lib/patches/db/mysql2.rb:8:in `each'                                                                                             
        6: from /var/lib/gems/2.5.0/gems/rack-mini-profiler-1.0.0/lib/patches/db/mysql2.rb:8:in `each'                                                                                             
        5: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/adapters/mysql2.rb:239:in `block (2 levels) in fetch_rows'                                                                        
        4: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/dataset/actions.rb:151:in `block in each'                                                                                         
        3: from /home/ubuntu/production/lib/cdo/contact_rollups.rb:814:in `block in update_data_from_forms'                                                                                        
        2: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/connection_pool/threaded.rb:91:in `hold'                                                                                          
        1: from /var/lib/gems/2.5.0/gems/sequel-5.9.0/lib/sequel/connection_pool/threaded.rb:91:in `abandon_results!'                                                                              
Sequel::DatabaseError (Mysql2::Error: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'souza@tdsb.on.ca';)   
    UPDATE `contact_rollups_daily` SET `street_address` = NUL' at line 9                                                                                                                           
[production] dashboard > 
```

**Comment by sureshc** [comment]
The `update_data_from_forms` method uses MULTI STATEMENTS to query rows based on the values of properties stored in the JSON `data` column.  Now that our database supports JSON functions, we may be able to refactor this method to eliminate the MULTI STATEMENTS.  Unfortunately, the set of JSON properties used in each row in the `forms` table varies wildly based on the form `kind` and the amount of fields filled out by the user.

@hacodeorg Determined that only two types of forms that are currently processed with multi-statements have had new data added recently.  This limits the set of JSON properties that we need to extract in SQL.

``` sql
SELECT DISTINCT JSON_KEYS(data)
FROM pegasus.forms
WHERE kind IN ('Petition', 'K5OnlineProfessionalDevelopmentPostSurvey')
ORDER BY id DESC;
```

```
["age_i", "name_s", "role_s", "email_s", "country_s"]
["age_i", "name_s", "role_s", "email_s", "country_s", "zip_code_s", "state_code_s"]
["name_s", "email_s", "school_s", "cscareers_i", "knowledge_i", "necessity_i", "confidence_i", "improve_pd_s", "teaching_cs_s", "satisfaction_s", "teacher_city_s", "teacher_type_s", "teacher_state_s", "teacher_years_s", "course_offering_s", "experience_time_i", "number_teaching_i", "teacher_country_s", "teacher_grades_ss", "experience_level_i", "teacher_describe_s", "experience_useful_i", "attend_in_person_pd_i", "attend_another_online_pd_i"]
["name_s", "email_s", "school_s", "cscareers_i", "knowledge_i", "necessity_i", "confidence_i", "improve_pd_s", "teaching_cs_s", "satisfaction_s", "teacher_city_s", "teacher_type_s", "teacher_state_s", "teacher_years_s", "course_offering_s", "experience_time_i", "number_teaching_i", "teacher_country_s", "teacher_grades_ss", "experience_level_i", "teacher_describe_s", "experience_useful_i", "teacher_type_other_s", "attend_in_person_pd_i", "attend_another_online_pd_i"]
["name_s", "email_s", "school_s", "cscareers_i", "knowledge_i", "necessity_i", "confidence_i", "improve_pd_s", "teaching_cs_s", "satisfaction_s", "teacher_city_s", "teacher_type_s", "teacher_state_s", "teacher_years_s", "course_offering_s", "experience_time_i", "number_teaching_i", "teacher_country_s", "teacher_grades_ss", "experience_level_i", "teacher_describe_s", "experience_useful_i", "attend_in_person_pd_i", "teacher_describe_other_s", "attend_another_online_pd_i"]
["name_s", "email_s", "school_s", "cscareers_i", "knowledge_i", "necessity_i", "confidence_i", "improve_pd_s", "teaching_cs_s", "satisfaction_s", "teacher_city_s", "teacher_type_s", "teacher_years_s", "course_offering_s", "experience_time_i", "number_teaching_i", "teacher_country_s", "teacher_grades_ss", "experience_level_i", "teacher_describe_s", "experience_useful_i", "attend_in_person_pd_i", "attend_another_online_pd_i"]
["name_s", "email_s", "school_s", "cscareers_i", "knowledge_i", "necessity_i", "confidence_i", "improve_pd_s", "teaching_cs_s", "satisfaction_s", "teacher_city_s", "teacher_type_s", "teacher_state_s", "teacher_years_s", "course_offering_s", "experience_time_i", "number_teaching_i", "teacher_country_s", "teacher_grades_ss", "experience_level_i", "teacher_describe_s", "experience_useful_i", "teacher_type_other_s", "attend_in_person_pd_i", "teacher_describe_other_s", "attend_another_online_pd_i"]
["name_s", "email_s", "school_s", "cscareers_i", "knowledge_i", "necessity_i", "confidence_i", "improve_pd_s", "teaching_cs_s", "satisfaction_s", "teacher_city_s", "teacher_type_s", "teacher_years_s", "course_offering_s", "experience_time_i", "number_teaching_i", "teacher_country_s", "teacher_grades_ss", "experience_level_i", "teacher_describe_s", "experience_useful_i", "teacher_type_other_s", "attend_in_person_pd_i", "teacher_describe_other_s", "attend_another_online_pd_i"]
["name_s", "email_s", "school_s", "cscareers_i", "knowledge_i", "necessity_i", "confidence_i", "improve_pd_s", "teaching_cs_s", "satisfaction_s", "teacher_city_s", "teacher_type_s", "teacher_years_s", "course_offering_s", "experience_time_i", "number_teaching_i", "teacher_country_s", "teacher_grades_ss", "experience_level_i", "teacher_describe_s", "experience_useful_i", "attend_in_person_pd_i", "teacher_describe_other_s", "attend_another_online_pd_i"]
["name_s", "email_s", "school_s", "cscareers_i", "knowledge_i", "necessity_i", "confidence_i", "improve_pd_s", "teaching_cs_s", "satisfaction_s", "teacher_city_s", "teacher_type_s", "teacher_years_s", "course_offering_s", "experience_time_i", "number_teaching_i", "teacher_country_s", "teacher_grades_ss", "experience_level_i", "teacher_describe_s", "experience_useful_i", "teacher_type_other_s", "attend_in_person_pd_i", "attend_another_online_pd_i"]
["age_i", "name_s", "role_s", "email_s", "zip_code_s"]
["name_s", "email_s"]
["name_s", "role_s", "email_s", "country_s"]
["name_s", "role_s", "email_s", "country_s", "zip_code_s", "state_code_s"]
```

**Comment by sureshc** [comment]
@hacodeorg Yesterday, I prototyped some raw SQL queries that extract data from JSON properties on the MySQL server using [JSON functions](https://dev.mysql.com/doc/refman/5.7/en/json.html):

``` sql
SELECT id
  ,            kind
  ,            data->>'$.email_s'                   AS email
  ,            NULL                                       AS street_address -- The Petition and K5OnlineProfessionalDevelopmentPostSurvey forms do not collect street address.
  ,            data->>'$.teacher_city_s'        AS city
  ,            data->>'$.state_code_s'          AS state_petition
  ,            data->>'$.teacher_state_s'      AS state_K5Survey
  ,            data->>'$.zip_code_s'             AS postal_code
  ,            data->>'$.country_s'                AS country_petition
  ,            data->>'$.teacher_country_s'  AS country_K5Survey
  ,            created_at
  ,            updated_at
FROM pegasus.forms
WHERE kind IN ('K5OnlineProfessionalDevelopmentPostSurvey')
ORDER BY id DESC
```

You can use the `Sequel.lit()` method in conjunction with MySQL JSON functions to extract properties from the `pegasus.forms.data` column:

`PEGASUS_REPORTING_DB_READER[:forms].select(Sequel.lit("id, JSON_EXTRACT(data, '$.email_s')"))`

**Comment by hacodeorg** [comment]
Creating temp table to test new code
```sql
create temporary table contact_rollups_daily_temp select * from contact_rollups_daily;
select count(*) from contact_rollups_daily_temp;
```

Combining all roles updating (from pegasus.forms) into 1 single statement
```sql
-- v2 (optimize for join)
UPDATE contact_rollups_daily_temp dest
INNER JOIN
  (SELECT * FROM forms WHERE forms.kind = 'Petition' and forms.created_at > '2019-05-01') AS petitions
ON dest.email = petitions.data->>'$.email_s'
SET dest.form_roles =
  CASE LOCATE(petitions.data->>'$.role_s', coalesce(dest.form_roles,''))
      when 0 then left(concat(coalesce(concat(dest.form_roles, ','), ''), petitions.data->>'$.role_s'),4096)
      else dest.form_roles
  END;

-- v1
UPDATE contact_rollups_daily_temp dest
INNER JOIN forms
ON dest.email = forms.data->>'$.email_s'
SET dest.form_roles =
  CASE LOCATE(forms.data->>'$.role_s', coalesce(dest.form_roles,''))
      when 0 then left(concat(coalesce(concat(dest.form_roles, ','), ''), forms.data->>'$.role_s'),4096)
      else dest.form_roles
  END
WHERE forms.kind = 'Petition' and forms.created_at > '2019-05-01';
```

**Comment by sureshc** [comment]
We identified some obstacles to using TEMPORARY tables, such as MySQL’s restriction that a temporary table cannot be referenced more than once in a query.  Many of the batch aggregation queries in Contact Rollups do this.  There may have been workarounds, but without unit tests on each query, it would be have been difficult to ensure that the modified queries were working correctly.

We adopted another approach #28767.

**Pull request closed without merging**